File: ruby/name_normalizer.rb
class NameNormalizer
  def normalize(name)
    return name if name == ""
    parse(name)
    return name if is_mononym?
    return "#{last}, #{first}" if is_duonym?
    return "#{last}, #{first} #{middle_initials}#{suffix}"
  end

  private
  def throw_on_excess_commas(name)
    raise ArgumentError.new if name.count(",") > 1
  end

  def parse(name)
    throw_on_excess_commas name
    @base_name, @suffix = name.split(",")
    @parts = @base_name.split(" ")
  end

  def is_mononym?
    @parts.size == 1
  end

  def is_duonym?
    @parts.size == 2
  end

  def initial(name_part)
    name_part.size == 1 ? name_part : "#{name_part[0]}."
  end

  def middle_initials
    @parts
      .slice(1..-2)
      .map {| part | initial(part)}
      .join(" ")
  end

  def suffix
    if @suffix then ",#{@suffix}" else "" end
  end

  def first
    @parts[0]
  end

  def last
    @parts[-1]
  end
end

File: typescript/src/nameNormalizer.ts
export function normalize(input: string): string {
  throwWhenExcessCommas(input);
  const trimmed = input.trim();
  if (isMononym(trimmed)) {
    return trimmed;
  }

  const nameParts = splitToNameParts(trimmed);
  if (isDuonym(nameParts)) {
    const { first, last } = nameParts;
    return `${last}, ${first}`;
  }
  return `${nameParts.last}, ${nameParts.first} ${formatMiddleInitials(
    nameParts.middles
  )}${formatSuffix(nameParts.suffix)}`;
}

const throwWhenExcessCommas = (input: string) => {
  const commaCount = [...input].filter((c) => c === ",").length;
  if (commaCount > 1) throw new Error();
};

const isMononym = (trimmed: string) => !trimmed.trim().includes(" ");

const isDuonym = (parts: NameParts) =>
  parts.middles.length === 0 && !parts.suffix;

const formatMiddleInitial = (middleName: string) =>
  middleName.length === 1 ? middleName : middleName[0] + ".";

const formatMiddleInitials = (parts: string[]) =>
  parts.map(formatMiddleInitial).join(" ");

const formatSuffix = (suffix?: string) => (suffix ? `, ${suffix}` : "");

type NameParts = {
  first: string;
  last: string;
  middles: string[];
  suffix?: string;
};

const splitToNameParts = (input: string): NameParts => {
  const [fullName, suffix] = input.split(", ");
  const nameParts = fullName.split(" ");
  return {
    first: nameParts[0],
    last: nameParts[nameParts.length - 1],
    middles: nameParts.slice(1, -1),
    suffix: suffix,
  };
};
